akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGroupBySpec.scala

```diff
@@ -643,7 +643,16 @@ class FlowGroupBySpec extends StreamSpec("""
           }
         }
       }
-      upstreamSubscription.sendComplete()
+
+      // complete has some try to feed downstream logic, we're not testing that here
+      // we just want to kill all substreams so lets fail the stream instead
+      upstreamSubscription.sendError(TE("killing this stream"))
+      map.values.foreach { subFlowState =>
+        // not all may have seen a subscription because random selection above
+        subFlowState.probe.ensureSubscription()
+        subFlowState.probe.expectError()
+
+      }
```

Comment:
This was the actual problem, the test would randomly cause buffered elements for the downstreams which means they wouldn't be completed until there was demand and therefore sometimes staying running after test was otherwise complete.